pkg/cert/source/reconciler.go

```diff
@@ -459,21 +455,9 @@ func (r *sourceReconciler) updateEntry(logger logger.LogContext, info CertInfo,
 			mod.Modify(true)
 		}
 
-		oldAlgorithm := ""
-		oldKeySize := 0
-		if spec.PrivateKey != nil {
-			oldAlgorithm = string(spec.PrivateKey.Algorithm)
-			oldKeySize = spec.PrivateKey.Size
-		}
-		if info.PrivateKeyAlgorithm != oldAlgorithm || info.PrivateKeySize != oldKeySize {
-			if info.PrivateKeyAlgorithm != "" || info.PrivateKeySize != 0 {
-				spec.PrivateKey = &api.CertificatePrivateKey{
-					Algorithm: api.PrivateKeyAlgorithm(info.PrivateKeyAlgorithm),
-					Size:      info.PrivateKeySize,
-				}
-			} else {
-				spec.PrivateKey = nil
-			}
+		newPrivateKey := createPrivateKey(info.PrivateKeyAlgorithm, info.PrivateKeySize)
+		if !reflect.DeepEqual(spec.PrivateKey, newPrivateKey) {
+			spec.PrivateKey = newPrivateKey
```

Comment:
Not sure if I understand the code correctly: Does this mean that the controller writes back the defaulted values to the `certificate.spec.privateKey` if it's not set before? If yes, won't this cause all existing certificates to be re-generated when this feature is rolled out because the spec checksum changes?

Reply:
If the source object has no private key related annotations, the function `createPrivateKey` returns `nil`. Even if you would add these annotations with the default values, the generated certificate resource will be updated, but no new certificate would be requested, as the hash code is unchanged.